evap/staff/views.py

```diff
@@ -462,6 +456,10 @@ def helper_course_edit(request, semester, course):
     if form.is_valid() and formset.is_valid():
         if operation not in ('save', 'approve'):
             raise SuspiciousOperation("Invalid POST operation")
+
+        if (not course.can_staff_edit or course.is_archived) and operation in ('save', 'approve'):
```

Comment:
the operation must be either `save` or `approve`at this point, so `if not course.can_staff_edit or course.is_archived:` would be enough.